File: frameworks/python/sanic/Implementation/routes/authorized.py
from functools import wraps

from sanic import Blueprint
from sanic.exceptions import Forbidden
from sanic.response import json

from documented import Payload, answers, refuses
from payloads import Payloads


def blueprint(p: Payloads) -> Blueprint:
    """authorized: Sanic has no authorization of its own, so the route carries the decorator Sanic's
    authentication guide writes, which checks the request before the handler runs."""
    token = p.settings["token"]

    # rb:wiring authorized.*
    def protected(handler):
        """Reads the bearer token Sanic parses from Authorization into request.token. Any token but
        settings.json's, or none, is Sanic's own Forbidden, 403."""

        @wraps(handler)
        async def checked(request, *args, **kwargs):
            if request.token != token:
                raise Forbidden()
            return await handler(request, *args, **kwargs)

        return checked
    # rb:end

    routes = Blueprint("authorized")

    @routes.get("/authorized/small")
    @answers(Payload, "items.small")
    @refuses(403, "Sanic's Forbidden, for any token but the accepted one")
    @protected
    async def small(request):
        return json(p.small)

    return routes

File: frameworks/node/express/Implementation/routes/authorized.ts
import type { RequestHandler } from "express";

import type { Routes } from "../app.ts";

// rb:wiring authorized.*
/**
 * The token check, as middleware in the route's handler list, which runs before the handler. A
 * token that is not settings.json's is refused with res.sendStatus(403), which writes the status's
 * name as a text/plain body.
 */
function requireToken(token: string): RequestHandler {
  const accepted = `Bearer ${token}`;
  return (request, response, next) => {
    if (request.headers.authorization === accepted) next();
    else response.sendStatus(403);
  };
}
// rb:end

/** authorized: Express has no authorization of its own, so the check is middleware on the route. */
const authorized: Routes = (app, p) => {
  app.get("/authorized/small", requireToken(p.settings.token), (_request, response) => response.json(p.small));
};

export default authorized;
